### PR TITLE
Added fix for RPC port detection

### PR DESCRIPTION
**Pull request by dat-boris**
Fixes: #5713
**Related**: #5590
**Merge before/after**: None

**Description**

In #5713 we discover that the new port detection logic does not apply
to GRPC startup, which default starts on Mac on IPv4 loopback.

Test plan:

Locally run multiple version of skaffold, ensure that none
of them have issues with starting up their own RPC server.

Instead of crashing, it now correctly skip previous ports used by RPC

```
WARN[0000] starting gRPC server on port 50053. (50051 is already in use) 
WARN[0000] starting gRPC HTTP server on port 50054. (50052 is already in use) 
```

### CODE REVIEW

**Comment by dat-boris** [comment]
This is my first PR for skaffold so any review comments are much appreciated.  I know also #5590 so would be nice if we can confirm that on Windows machine this will not cause any regression (hopefully shouldn't do - it was a revert to previous behaviour for GRPC).

File: pkg/skaffold/util/port.go [hunk 1]
```diff
@@ -87,37 +90,39 @@ func (f *PortSet) List() []int {
...
 func GetAvailablePort(port int, usedPorts *PortSet) int {
-	if port > 0 {
-		if getPortIfAvailable(port, usedPorts) {
-			return port
-		}
+	// We map this to allow
+	return GetAvailablePortWithAddreess(Any, port, usedPorts)
+}
+
+func GetAvailablePortWithAddreess(address string, port int, usedPorts *PortSet) int {
```
Comment (MarlonGamez) [inline pkg/skaffold/util/port.go:104]:
```suggestion
func GetAvailablePortWithAddress(address string, port int, usedPorts *PortSet) int {
```

File: pkg/skaffold/util/port.go [hunk 2]
```diff
@@ -87,37 +90,39 @@ func (f *PortSet) List() []int {
...
 // First, check if the provided port is available on the specified address. If so, use it.
 // If not, check if any of the next 10 subsequent ports are available.
 // If not, check if any of ports 4503-4533 are available.
 // If not, return a random port, which hopefully won't collide with any future containers
-//
-// See https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt
+
+// See https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt,
 func GetAvailablePort(port int, usedPorts *PortSet) int {
-	if port > 0 {
```
Comment (MarlonGamez) [inline pkg/skaffold/util/port.go:98]: do we want to get rid of this check?

File: pkg/skaffold/util/port.go [hunk 3]
```diff
@@ -87,37 +90,39 @@ func (f *PortSet) List() []int {
...
 
-// GetAvailablePort returns an available port that is near the requested port when possible.
 // First, check if the provided port is available on the specified address. If so, use it.
 // If not, check if any of the next 10 subsequent ports are available.
 // If not, check if any of ports 4503-4533 are available.
 // If not, return a random port, which hopefully won't collide with any future containers
-//
-// See https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt
+
+// See https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt,
```
Comment (MarlonGamez) [inline pkg/skaffold/util/port.go:98]:
```suggestion
// See https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.txt
```

**Review by MarlonGamez — COMMENTED** [review]
Thanks for opening this @dat-boris ! Have a few tiny nit comments :)

File: pkg/skaffold/util/port.go (hunk 2 above)
Comment (dat-boris) [reply pkg/skaffold/util/port.go:98]: Good catch - thank you!

File: pkg/skaffold/util/port.go [hunk 4]
```diff
@@ -129,16 +138,20 @@ func GetAvailablePort(port int, usedPorts *PortSet) int {
...
 func IsPortFree(p int) bool {
-	l, err := net.Listen("tcp", fmt.Sprintf(":%d", p))
+	return IsPortFreeWithAddress(Any, p)
+}
+
+func IsPortFreeWithAddress(address string, p int) bool {
+	l, err := net.Listen("tcp", fmt.Sprintf("%s:%d", address, p))
 	if err != nil {
 		return false
 	}
```
Comment (briandealwis) [inline pkg/skaffold/util/port.go:157]: So this fix isn't quite enough as it still leaves the previous situation, where a local port may be available but the INADDR_ANY:port isn't available.  We should test both conditions.   Here's my suggestion:
```suggestion
	// Ensure the port is available across all interfaces
	if l, err := net.Listen("tcp", fmt.Sprintf(":%d", p));  err != nil {
		return false
	} else {
		l.Close()
	}
	if address != Any {
		// Ensure the port is available on the specific interface too
		if l, err := net.Listen("tcp", fmt.Sprintf("%s:%d", address, p));  err != nil {
			return false
		} else {
			l.Close()
		}
	}
```

**Comment by briandealwis** [comment]
Sorry @dat-boris for the delay — I held off as I wanted to check how we were using `GetAvailablePort()`. Prior to #5670 we were explicitly passing the interface.  I think we should restore that aspect and I'll push that change here.

**Review by MarlonGamez — APPROVED** [review]
lgtm, thanks @dat-boris for the contribution :)

**Comment by dat-boris** [comment]
Thank you @briandealwis @MarlonGamez for the reviews! Yeah it make sense to revert to the old `GetAvailablePort` and improve the check!

**Pull request merged**